### PR TITLE
sys/sema: add sema_ztimer64 to implement old api, deprecate sema

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR takes for `sema` a similar approach to `event_timeout`. The module is switched to `ztimer` removing the `xtimer` based implementation.

To keep compatibility the old api (using 64bit timeouts) is kept if selecting `sema` module, which uses `ztimer64`, but this module is marked now as deprecated.

This PR also migrates posix to use `ztimer64` and in therefore gives a user for `sema_ztimer64` by using it to implement `posix_sempahore`

### Testing procedure

- green murdock

### CODE REVIEW

**Comment by fjmolinas** [comment]
@kfessel I updated this PR to take care of the sema issues.

**Comment by fjmolinas** [comment]
All green except for an unrelated test failure:

```
--- static tests: passed
-- running on worker riotbuild-0 thread 3, build number 50627.
--- can_fast_ci_run:
General build system / KConfig changes require a full CI run

-- can_fast_ci_run.py exited non-zero

---
--- compile job results (1 failed, 107605 passed, 107606 total):
bootloaders/riotboot (96/96):
```

```
-- running on worker mobi7 thread 11, build number 55330.
make: Entering directory '/tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/tests/driver_si114x'
rm -rf /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/pkg-build/esp8266_sdk
make: Leaving directory '/tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/tests/driver_si114x'
make: Entering directory '/tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/tests/driver_si114x'
[INFO] cloning esp8266_sdk
Building application "tests_driver_si114x" for "esp8266-esp-12x" with MCU "esp8266".

git-cache: cloning from cache. tag=commit913a06a9ac3b2f18009e8fee8f092ca9ffeccd38-495945
[INFO] updating esp8266_sdk /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/pkg/esp8266_sdk/.pkg-state.git-downloaded
echo 913a06a9ac3b2f18009e8fee8f092ca9ffeccd38 > /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/pkg/esp8266_sdk/.pkg-state.git-downloaded
[INFO] patch esp8266_sdk
   text	   data	    bss	    dec	    hex	filename
 281732	   5312	  28760	 315804	  4d19c	/tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/tests_driver_si114x.elf
esptool.py v2.4.0
Parsing CSV input...
sha1sum /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/tests_driver_si114x.elf /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/tests_driver_si114x.elf.bin /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/partitions.bin > /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/test-input-hash.sha1
make: Leaving directory '/tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/tests/driver_si114x'
make: Entering directory '/tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/tests/driver_si114x'
Building application "tests_driver_si114x" for "esp8266-esp-12x" with MCU "esp8266".
rm -rf /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/pkg-build/esp8266_sdk

[INFO] cloning esp8266_sdk
git-cache: cloning from cache. tag=commit913a06a9ac3b2f18009e8fee8f092ca9ffeccd38-517399
[INFO] updating esp8266_sdk /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/pkg/esp8266_sdk/.pkg-state.git-downloaded
echo 913a06a9ac3b2f18009e8fee8f092ca9ffeccd38 > /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/pkg/esp8266_sdk/.pkg-state.git-downloaded
[INFO] patch esp8266_sdk
   text	   data	    bss	    dec	    hex	filename
 281956	   5312	  28760	 316028	  4d27c	/tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/tests_driver_si114x.elf
esptool.py v2.4.0
Parsing CSV input...
sha1sum /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/tests_driver_si114x.elf /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/tests_driver_si114x.elf.bin /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/partitions.bin > /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/test-input-hash.sha1
make: Leaving directory '/tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/tests/driver_si114x'
Hashes of binaries with and without Kconfig mismatch for tests/driver_si114x with esp8266-esp-12x
Please check that all used modules are modelled in Kconfig and enabled
rm -rf /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/pkg-build/esp8266_sdk
rm -rf /tmp/dwq.0.4159310719130117/e1aac77a25f50fef10db80e94c331706/build/pkg-build/esp8266_sdk
-- build directory size: 100K
```

**Comment by fjmolinas** [comment]
Unless there are comments I think it would be Ok to skip compile test in next build.

**Review by kfessel — COMMENTED** [review]
how to use the new sema without any timer?

**Comment by fjmolinas** [comment]
> how to use the new sema without any timer?

That's a good question, but it's a bit out of the scope of this PR. Currently, you need a timer for `sema`, there is an explicit dependency on either `xtimer` or `ztimer` being selected. So this PR does not change this, it just removes the `xtimer` part.

I think it would be good not to depend on this, but again it's out of scope.

**Comment by kaspar030** [comment]
I think we need a stance on renaming modules. If we give modules a new name for breaking changes like this, "sema" users of the future will need to be educated to use "sema_ztimer64" instead of "sema", right?

Wouldn't we rather want the current up-to-date version to be "sema", and the deprecated one "sema_something"?

I think it is fine to ask users to update a dependency name when upgrading RIOT to a new release, *if it is mentioned in the release notes*.

**Comment by kfessel** [comment]
> That's a good question, but it's a bit out of the scope of this PR. Currently, you need a timer for `sema`, there is an explicit dependency on either `xtimer` or `ztimer` being selected. So this PR does not change this, it just removes the `xtimer` part.
> 
> I think it would be good not to depend on this, but again it's out of scope.

No, `sema` has no deppendacy on anything (just core and thats a given).

**Comment by fjmolinas** [comment]
> I think we need a stance on renaming modules. If we give modules a new name for breaking changes like this, "sema" users of the future will need to be educated to use "sema_ztimer64" instead of "sema", right?
> 
> Wouldn't we rather want the current up-to-date version to be "sema", and the deprecated one "sema_something"?
> 
> I think it is fine to ask users to update a dependency name when upgrading RIOT to a new release, _if it is mentioned in the release notes_.

So instead have `sema`, `sema_64` and `sema_legacy` modules, something like that?

**Comment by kaspar030** [comment]
> So instead have `sema`, `sema_64` and `sema_legacy` modules, something like that?

I'd go with just `sema` (new ztimer one where a clock needs to be selected) and `sema_deprecated` (old API that maps to ztimer64).

If a clock can be specified, we don't need to support 64bit timeouts. Or does a use-case come to mind?

**Comment by fjmolinas** [comment]
> > So instead have `sema`, `sema_64` and `sema_legacy` modules, something like that?
> 
> I'd go with just `sema` (new ztimer one where a clock needs to be selected) and `sema_deprecated` (old API that maps to ztimer64).
> 
> If a clock can be specified, we don't need to support 64bit timeouts. Or does a use-case come to mind?

Posix semaphores use 64bits AFAIKl, so changing the clock is not enough since we can't really mix clock backends safely here.

**Comment by kaspar030** [comment]
> Posix semaphores use 64bits AFAIKl,

yes, just [checked](https://linux.die.net/man/3/sem_wait), it allows to specify an absolute timeout using a `struct timespec`.
so until that's sorted out, can't we still go with just "sema" and "sema_deprecated", and make posix use the latter for now?
down the line, we fix the posix time story and provide it with proper epoch, maybe ms based?

**Comment by fjmolinas** [comment]
> yes, just [checked](https://linux.die.net/man/3/sem_wait), it allows to specify an absolute timeout using a `struct timespec`.
> so until that's sorted out, can't we still go with just "sema" and "sema_deprecated", and make posix use the latter for now?
> down the line, we fix the posix time story and provide it with proper epoch, maybe ms based?

OK!

File: sys/include/sema.h [hunk 1]
```diff
@@ -191,16 +187,39 @@ static inline int sema_wait(sema_t *sema)
...
+ * @return  -ECANCELED, if the semaphore was destroyed.
+ * @return  -EAGAIN,    if the semaphore is not posted (only if timeout = 0)
+ */
+static inline int sema_wait_timed_ztimer64(sema_t *sema,
+                                           ztimer64_clock_t *clock,
+                                           uint32_t timeout)
+{
+    return _sema_wait_ztimer64(sema, (timeout != 0), clock, timeout);
+}
+#endif
```
Comment (kfessel) [inline sys/include/sema.h:216]: i don't think this api should be created or it should be marked internal use only.

(i somehow can not imagine a case where someone wants to wait for a semaphorer for a time longer than 1h but not infinite) 

1h ~ fits into 32Bit of usec or 22Bit of msec

Comment (kfessel) [reply sys/include/sema.h:216]: (there certainly is the case of posix semaphore , and i found a similar one in wasm but in both cases it seems to me that the person who designed that api just used the time type that allready was there)

Comment (fjmolinas) [reply sys/include/sema.h:216]: Ok so drop this from the public API!

**Comment by fjmolinas** [comment]
> Wouldn't we rather want the current up-to-date version to be "sema", and the deprecated one "sema_something"?

Renamed the modules, will need to squash...

**Comment by fjmolinas** [comment]
@kaspar030 If the next build is green, I would suggest not triggering a full-build again, if after squashing only the commit message change.

**Comment by fjmolinas** [comment]
All green! @kfessel @kaspar030 can I squash? Should we skip the build?

**Comment by fjmolinas** [comment]
> All green! @kfessel @kaspar030 can I squash? Should we skip the build?

ups I didn't mean to comment this yet..

**Comment by fjmolinas** [comment]
> > All green! @kfessel @kaspar030 can I squash? Should we skip the build?
> 
> ups I didn't mean to comment this yet..

Ok now yes :D

**Review by kaspar030 — APPROVED** [review]
ACK.

**Comment by fjmolinas** [comment]
Squashed just changing commit message.

**Comment by kfessel** [comment]
added that skip to late

**Pull request merged**